nexus/db-queries/src/db/datastore/disk.rs

```diff
@@ -208,6 +324,50 @@ impl DataStore {
 
                 Disk::Crucible(CrucibleDisk { disk, disk_type_crucible })
             }
+
+            db::model::DiskType::LocalStorage => {
+                use nexus_db_schema::schema::disk_type_local_storage::dsl;
+
+                let disk_type_local_storage = dsl::disk_type_local_storage
+                    .filter(dsl::disk_id.eq(disk_id))
+                    .select(DiskTypeLocalStorage::as_select())
+                    .first_async(&*conn)
+                    .await
+                    .map_err(|e| {
+                        public_error_from_diesel(e, ErrorHandler::Server)
+                    })?;
+
+                let local_storage_dataset_allocation = if let Some(
+                    allocation_id,
+                ) =
+                    disk_type_local_storage
+                        .local_storage_dataset_allocation_id()
+                {
+                    use nexus_db_schema::schema::local_storage_dataset_allocation::dsl;
+
+                    Some(
+                        dsl::local_storage_dataset_allocation
+                            .filter(dsl::id.eq(to_db_typed_uuid(allocation_id)))
+                            .select(LocalStorageDatasetAllocation::as_select())
+                            .first_async(&*conn)
+                            .await
+                            .map_err(|e| {
+                                public_error_from_diesel(
+                                    e,
+                                    ErrorHandler::Server,
+                                )
+                            })?,
```

Comment:
might we want to add an `internal_context` to these errors saying which table didn't have an entry for this disk? if we encountered a case where a `disk` record with the local storage disk type exists, but the `disk_type_local_storage` record for it was gone, or the disk had an allocation ID but the record in the allocation table was not there, I think it would be helpful if the logs indicated _which_ record was missing that should exist.